### ISSUE TITLE
sd_bus vtable methods do not have offset member

### ISSUE THREAD

**Issue by daurnimator** [systemd/systemd#577]
If you use `_SD_BUS_VTABLE_METHOD` you have no way to pass a custom `userdata` per handler.

With a property (e.g. `_SD_BUS_VTABLE_PROPERTY`) you have the `offset` member; but methods have nothing.

**Comment by dvdhrm** [comment]
You really should not misuse `offset` for userdata. This might have nasty side-effects as we use it for builtin handlers. Instead, you have to pass the userdata via `sd_bus_add_object_vtable()`, or, in case of fallback vtables, your `find` callback in `sd_bus_add_fallback_vtable()` should return the userdata that is to be passed to the method call.

**Comment by daurnimator** [comment]
> You really should not misuse offset for userdata.

This is exactly what `offset` is for => the userdata passed to property handlers is the vtable's userdata + the offset.

The issue is that methods don't have the same allowance.

**Comment by dvdhrm** [comment]
No. The `offset` field is used to avoid passing a `get` / `set` callback. If those a `NULL`, the `offset` is used to directly store / retrieve the data from your passed object. This is only a fast path to avoid hard-coding the same handlers over and over again. The fallback is always `offset == NULL`.

In other words, whenever a vtable callback is called, it is called on an object path. This object path belongs to an object. This object must have either been registered before via `sd_bus_add_*foobar*()`, in which case you passed userdata. Or it must be a fallback, in which case your lookup function provides the userdata.

There is no separate userdata for each vtable-entry, as this is not how this is supposed to be used. Instead, your callback _knows_ which entry it belongs to. All it needs is the context to know the object it is called on.

Again, you really should not mis-use the `offset` as userdata. Instead, provide a different `get` / `set` function for each property, if you need such context.

**Comment by daurnimator** [comment]
> Instead, your callback knows which entry it belongs to.

I'm using this in a dynamic context: I have only one C callback that all functions use; but they dispatch to the correct method by tracking this in the userdata.

> No. The offset field is used to avoid passing a get / set callback. If those a NULL, the offset is used to directly store / retrieve the data from your passed object. 

Really? how does it know the type + memory layout of my object?

> Again, you really should not mis-use the offset as userdata. Instead, provide a different get / set function for each property, if you need such context.

As I noted; this is how it works for `get`/`set` callbacks today; and it works well for my usecase.

**Comment by dvdhrm** [comment]
> I'm using this in a dynamic context: I have only one C callback that all functions use; but they dispatch to the correct method by tracking this in the userdata.

I don't think this is a good idea. Anyway, you're still free to access `sd_bus_message_get_member()` and `sd_bus_message_get_signature()` to figure out which function was called.

> Really? how does it know the type + memory layout of my object?

It only works for basic types. It then assumes, that such basic type is stored at the passed offset.

**Comment by daurnimator** [comment]
> I don't think this is a good idea.

I don't have much of an alternative... unless I allocate pieces of executable memory to use as unique trampolines. This sounds very ugly.

> I don't think this is a good idea. Anyway, you're still free to access sd_bus_message_get_member() and sd_bus_message_get_signature() to figure out which function was called.

At that point I might as well just use `add_match` and write my whole own vtable-like thing....

But, IMO an offset member for methods makes sense.
It'll even be backwards compatible (existing code will have it set to `0`)

**Comment by daurnimator** [comment]
Example of my current (working) usage:

``` C
static int bus_property_set(sd_bus *bus, const char *path, const char *interface, const char *property, sd_bus_message *value, void *userdata, sd_bus_error *ret_error) {
    sd_bus_slot *slot = sd_bus_get_current_slot(bus);
    void *vtable_udata = sd_bus_slot_get_userdata(slot);
    size_t offset = userdata-vtable_udata;
```

I then use `offset` as a key to a hashtable accessed via `vtable_udata`.
This lets me find the relevant property handler.

**Comment by dvdhrm** [comment]
Why not use `sd_bus_add_object()` and avoid the whole vtable? I mean, you clearly use it a different way than it was intended to be used. You probably also use non-static vtables (otherwise, there'd be no reason to use the same callback function for all entries).

I also don't get why you do all these workarounds. Is this to write some kind of language binding?

**Comment by daurnimator** [comment]
> Why not use sd_bus_add_object() and avoid the whole vtable?

Because then I don't get the nicely auto generated introspection stuff.

> You probably also use non-static vtables (otherwise, there'd be no reason to use the same callback function for all entries).

Yes.

> I also don't get why you do all these workarounds.

So that I don't have to write my own dbus library (I've already done that twice; and I've never been happy with the results) and/or dispatching.

> Is this to write some kind of language binding?

Yes. https://github.com/daurnimator/lua-systemd/commits/sd-bus

**Comment by dvdhrm** [comment]
So, I'm no big fan of providing a toolkit for language bindings, so I don't see a point in this feature. However, I don't want to be the ultimate authority on that decision, so I'll leave this open for someone else to comment.

**Comment by poettering** [comment]
Hmm, I find this pretty crazy, in particular dynamically assembled vtables... That said, there's an asymmetry here indeed: we allow the userdata to be manipulated in more ways for properties than for methods. Of course, we never needed this for methods, and I find the usecase pretty weak, but it is admittedly an asymmetry.

I figure we could introduce an offset field without breaking compat, just for completeness sake... Dunno. 

If you prep a convincing patch I might be open to it... Dunno. 

Not enthusiastic about the idea... 

The offset is really just supposed to be an offset, and not more than that, just to allow generic implementations of property getters and setters for many fields in the same objects. One could probably make the weak case that generic method implementations might make sense too, that apply to different fields of the same object. Example: an object exposing two variables "Alpha" and "Beta" plus two methods "IncrementAlpha()" and "IncrementBeta()", that could be backed by a struct of two integers, plus a single method function "increment_integer()", that is then bound to both variables via the offset field if you follow what i mean. Such a usecase I still don't find too convincing, but also not completely wrong. Hence I would likely accept a patch that adds an offset parameter for method vtable entries if it does this cleanly, correctly, and does not break API nor ABI.

**Comment by daurnimator** [comment]
> If you prep a convincing patch I might be open to it...

Would it be as simple as:

``` diff
diff --git a/src/libsystemd/sd-bus/bus-objects.c b/src/libsystemd/sd-bus/bus-objects.c
index cbdf655..a973bca 100644
--- a/src/libsystemd/sd-bus/bus-objects.c
+++ b/src/libsystemd/sd-bus/bus-objects.c
@@ -68,6 +68,12 @@ static int node_vtable_get_userdata(
         return 1;
 }

+static void *vtable_method_convert_userdata(const sd_bus_vtable *p, void *u) {
+        assert(p);
+
+        return (uint8_t*) u + p->x.method.offset;
+}
+
 static void *vtable_property_convert_userdata(const sd_bus_vtable *p, void *u) {
         assert(p);

@@ -360,6 +366,8 @@ static int method_callbacks_run(
         if (bus->nodes_modified)
                 return 0;

+        u = vtable_method_convert_userdata(c->vtable, u);
+
         *found_object = true;

         if (c->last_iteration == bus->iteration_counter)
diff --git a/src/systemd/sd-bus-vtable.h b/src/systemd/sd-bus-vtable.h
index a032215..fcff6f5 100644
--- a/src/systemd/sd-bus-vtable.h
+++ b/src/systemd/sd-bus-vtable.h
@@ -67,6 +67,7 @@ struct sd_bus_vtable {
                         const char *signature;
                         const char *result;
                         sd_bus_message_handler_t handler;
+                        size_t offset;
                 } method;
                 struct {
                         const char *member;
@@ -97,6 +98,7 @@ struct sd_bus_vtable {
                 .x.method.signature = _signature,                       \
                 .x.method.result = _result,                             \
                 .x.method.handler = _handler,                           \
+                .x.method.offset = 0,                                   \
         }

 #define SD_BUS_SIGNAL(_member, _signature, _flags)                      \
```

I wasn't sure how you want to preserve compatibility with the current `SD_BUS_METHOD` macro signature, so I left that out for now.

**Comment by daurnimator** [comment]
For those following by email, I updated the patch above.

**Comment by poettering** [comment]
There should be a macro SD_BUS_VTABLE_METHOD_WITH_OFFSET() that is like SD_BUS_VTABLE_METHOD() but adds an offset initializer. Also, file as PR, please, and add fixes #577.
